README.MD
# HipChat Notification Resource

Send notification messages to HipChat using a string message or templated message.

## Building
RUN:
```
$ ./src/ci/build.sh
$ docker build -t dockerhub-username/dockerhub-repository .
$ docker push dockerhub-username/dockerhub-repository
```

## Installing

```
resource_types:
- name: hipchat-notification-resource
  type: docker-image
  source:
    repository: dockerhub-username/dockerhub-repository
```

## Source Configuration

* `hipchat_server_url`: *Required.* https://api.hipchat.com OR https://api.hipchat.com/v2/room/12456 (see room_id below)
* `token`: *Required.* token to authenticate with HipChat server
* `room_id`: The room to send notifications to (Required when not specified in the hipchat_server_url)

#### `out`: Sends message to HipChat room

Send message to specified HipChat Room, with the configured parameters

#### Parameters
* `from`: *Required.* Name of the system sending the notification
* `message`: *Required.* (string) Text to send to the HipChat room
* `message`: *Required.* (object)
  + `template` *Required.* Template string with values to replace from params in the format ${param_key}
  + `params` *Required.* Key and values to use to replace in template. Can start with file:// to get a value from a file. By default includes environment variables available in: https://concourse.ci/implementing-resources.html#resource-metadata
* `color`: *Optional.* One of "yellow", "red", "green", "purple", "gray", or "random". (default: yellow)
* `message_format`: *Optional.* Message format, either html or text (default: html)
* `notify`: *Optional.* If the message should trigger a notification to people in the room. 0 = false, 1 = true (default: 0)

## Example

### Check
```
---
resources:
- name: hipchat-notification
  type: hipchat-notification-resource
  source:
    hipchat_server_url: https://api.hipchat.com
    token: 14376e01-3152-4710-a830-aaf77e48c047
    room_id: 1
```

### In

*Not supported*

### Out

```
---
- put: hipchat-notification
  params:
    color: green
    from: "Concourse CI"
    message:
      template: "${PR-TITLE} Build Successful - <a href='${ATC_EXTERNAL_URL}/pipelines/${BUILD_PIPELINE_NAME}/jobs/${BUILD_JOB_NAME}/builds/${BUILD_NAME}'>Build Logs</a>"
      params:
        PR-TITLE: file://output-directory/pr-title
```


ci/release_notes.md
Moved to cloudfoundry-community/cfcommunity


src/ci/test.sh
#!/usr/bin/env bash

# (c) Copyright 2016 Hewlett Packard Enterprise Development LP
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

set -e

cd scripts
./node_modules/gulp/bin/gulp.js lint-and-beautify --fail-on-beautify true
./node_modules/mocha/bin/mocha tests/*Tests.js
cd ..


scripts/tests/hipChatTests.js
// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var should = require('should'),
  rewire = require('rewire'),
  sinon = require('sinon'),
  HipChatNotifier = rewire('../hipChatClient');

describe('hipchat notifier', function (done) {

  it("Should check for argument value", function () {
    var mockConsole = {
      error: sinon.stub()
    };

    HipChatNotifier.__set__({
      'console': mockConsole
    });

    new HipChatNotifier().checkArgument('name');
    mockConsole.error.calledWith('Please provide a value for', 'name')
      .should.equal(true);
  });

  it("Should exit(1) if no source variables are provided", function () {
    var mockConsole = {
        error: sinon.stub(),
        log: sinon.stub()
      },
      mockProces = {
        'exit': sinon.stub()
      };

    HipChatNotifier.__set__({
      'console': mockConsole,
      'process': mockProces
    });

    new HipChatNotifier().run({}, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });
    mockProces.exit.calledWith(1).should.equal(true);
  });

  it("Should exit(1) if no params variables are provided", function () {
    var mockConsole = {
        error: sinon.stub(),
        log: sinon.stub()
      },
      mockProces = {
        'exit': sinon.stub()
      };

    HipChatNotifier.__set__({
      'console': mockConsole,
      'process': mockProces
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com',
      'token': 'token',
      'room_id': 12456
    }, {});
    mockProces.exit.calledWith(1).should.equal(true);
  });


  it("Should exit(0) if room_id is in url", function () {
    var mockConsole = {
        error: sinon.stub().throws('Should not be called.'),
        log: sinon.stub()
      },
      mockProcess = {
        'exit': sinon.stub()
      },
      mockRequest = sinon.stub().callsArgWith(1, null, {
        statusCode: 200
      });

    HipChatNotifier.__set__({
      console: mockConsole,
      process: mockProcess,
      request: mockRequest
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com/v2/room/12456',
      'token': 'token',
    }, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });

    mockRequest.calledWith({
      url: 'https://api.hipchat.com/v2/room/12456/notification?auth_token=token',
      method: 'POST',
      json: {
        room_id: 12456,
        from: 'Concourse CI',
        message: 'Building',
        color: 'Yellow'
      }
    }).should.equal(true);

    mockProcess.exit.calledWith(0).should.equal(true);
  });

  it("Should exit(0) if all environment variables are provided", function () {
    var mockConsole = {
        error: sinon.stub().throws('Should not be called.'),
        log: sinon.stub()
      },
      mockProcess = {
        'exit': sinon.stub()
      },
      mockRequest = sinon.stub().callsArgWith(1, null, {
        statusCode: 200
      });

    HipChatNotifier.__set__({
      console: mockConsole,
      process: mockProcess,
      request: mockRequest
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com',
      'token': 'token',
      'room_id': 12456
    }, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });
    mockRequest.calledWith({
      url: 'https://api.hipchat.com/v2/room/12456/notification?auth_token=token',
      method: 'POST',
      json: {
        room_id: 12456,
        from: 'Concourse CI',
        message: 'Building',
        color: 'Yellow'
      }
    }).should.equal(true);

    mockProcess.exit.calledWith(0).should.equal(true);
  });

  it("Should exit(1) if room id is missing", function () {
    var mockConsole = {
        error: sinon.stub(),
        log: sinon.stub()
      },
      mockProcess = {
        'exit': sinon.stub()
      },
      mockRequest = sinon.stub().callsArgWith(1, null, {
        statusCode: 200
      });

    HipChatNotifier.__set__({
      console: mockConsole,
      process: mockProcess,
      request: mockRequest
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com',
      'token': 'token'
    }, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });

    mockProcess.exit.calledWith(1).should.equal(true);
  });

  it("Should exit(1) if response code is not 200", function () {
    var mockConsole = {
        error: sinon.stub(),
        log: sinon.stub()
      },
      mockProcess = {
        'exit': sinon.stub()
      },
      mockRequest = sinon.stub().callsArgWith(1, null, {
        statusCode: 400,
        body: "BAD REQUEST"
      });

    HipChatNotifier.__set__({
      console: mockConsole,
      process: mockProcess,
      request: mockRequest
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com',
      'token': 'token',
      'room_id': 12456
    }, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });
    mockRequest.calledWith({
      url: 'https://api.hipchat.com/v2/room/12456/notification?auth_token=token',
      method: 'POST',
      json: {
        room_id: 12456,
        from: 'Concourse CI',
        message: 'Building',
        color: 'Yellow'
      }
    }).should.equal(true);

    mockProcess.exit.calledWith(1).should.equal(true);
  });

  it("Should exit(0) if response code is not 200 and fail_on_error is false", function () {
    var mockConsole = {
        error: sinon.stub(),
        log: sinon.stub()
      },
      mockProcess = {
        'exit': sinon.stub()
      },
      mockRequest = sinon.stub().callsArgWith(1, null, {
        statusCode: 400,
        body: "BAD REQUEST"
      });

    HipChatNotifier.__set__({
      console: mockConsole,
      process: mockProcess,
      request: mockRequest
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com',
      'token': 'token',
      'room_id': 12456,
      'fail_on_error': false
    }, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });
    mockRequest.calledWith({
      url: 'https://api.hipchat.com/v2/room/12456/notification?auth_token=token',
      method: 'POST',
      json: {
        room_id: 12456,
        from: 'Concourse CI',
        message: 'Building',
        color: 'Yellow'
      }
    }).should.equal(true);

    mockProcess.exit.calledWith(0).should.equal(true);
  });

  it("Should exit(1) if response code is not 200 and fail_on_error is undefined", function () {
    var mockConsole = {
        error: sinon.stub(),
        log: sinon.stub()
      },
      mockProcess = {
        'exit': sinon.stub()
      },
      mockRequest = sinon.stub().callsArgWith(1, null, {
        statusCode: 400,
        body: "BAD REQUEST"
      });

    HipChatNotifier.__set__({
      console: mockConsole,
      process: mockProcess,
      request: mockRequest
    });

    new HipChatNotifier().run({
      'hipchat_server_url': 'https://api.hipchat.com',
      'token': 'token',
      'room_id': 12456
    }, {
      'message': 'Building',
      'from': 'Concourse CI',
      'color': 'Yellow'
    });
    mockRequest.calledWith({
      url: 'https://api.hipchat.com/v2/room/12456/notification?auth_token=token',
      method: 'POST',
      json: {
        room_id: 12456,
        from: 'Concourse CI',
        message: 'Building',
        color: 'Yellow'
      }
    }).should.equal(true);

    mockProcess.exit.calledWith(1).should.equal(true);
  });


});

scripts/tests/templateTests.js
// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var should = require('should'),
  rewire = require('rewire'),
  sinon = require('sinon'),
  tokenReplacer = rewire('../tokenReplacement');

describe('Notifier template', function (done) {

  it("Should replace a build token", function () {
    var mockProcess = {
        env: {
          BUILD_ID: 12
        }
      },
      message = {},
      params = null;

    message.template = "Test message with ${BUILD_ID}",

      tokenReplacer.__set__({
        'process': mockProcess
      });

    tokenReplacer.replaceTokens(message, params, function (error, newMessage) {
      newMessage.should.equal("Test message with 12");
    });
  });

  it("Should replace all instances of the same build token", function () {
    var mockProcess = {
      env: {
        BUILD_ID: 12,
        BUILD_JOB_NAME: 'Super-Job'
      }
    };

    tokenReplacer.__set__({
      'process': mockProcess
    });

    var message = {};
    message.template = "${BUILD_ID} Job Name: ${BUILD_JOB_NAME} Build Id: ${BUILD_ID}";
    var params = null;

    tokenReplacer.replaceTokens(message, params, function (error, newMessage) {
      newMessage.should.equal("12 Job Name: Super-Job Build Id: 12");
    });
  });

  it("Should replace all instances of a param token", function () {
    var mockProcess = {
        env: {
          BUILD_ID: 12,
          BUILD_JOB_NAME: 'Super-Job'
        }
      },
      message = {},
      params = null;

    tokenReplacer.__set__({
      'process': mockProcess
    });

    message.template = "${PR_URL} Pull Request # ${PR_NUMBER}";
    message.params = {
      PR_URL: 'http://www.google.com/',
      PR_NUMBER: '1'
    };

    tokenReplacer.replaceTokens(message, params, function (error, newMessage) {
      newMessage.should.equal("http://www.google.com/ Pull Request # 1");
    });
  });

  it("Should replace all instances of the same param token", function () {
    var mockProcess = {
        env: {
          BUILD_ID: 12,
          BUILD_JOB_NAME: 'Super-Job'
        }
      },
      message = {},
      params = null;

    tokenReplacer.__set__({
      'process': mockProcess
    });

    message.template = "PR: ${PR_NUMBER} - ${PR_URL} Pull Request # ${PR_NUMBER}";
    message.params = {
      PR_URL: 'http://www.google.com/',
      PR_NUMBER: '1'
    };

    tokenReplacer.replaceTokens(message, params, function (error, newMessage) {
      newMessage.should.equal("PR: 1 - http://www.google.com/ Pull Request # 1");
    });
  });


});

scripts/out.js
#! /usr/bin/node

// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var HipChatClient = require('./hipChatClient'),
  tokenReplacer = require('./tokenReplacement'),
  stdin = process.stdin,
  inputChunks = [];

stdin.setEncoding('utf8');

stdin.on('data', function (chunk) {
  inputChunks.push(chunk);
});

stdin.on('end', function () {
  var inputJSON = inputChunks.join("");

  var parsedData = JSON.parse(inputJSON.toString()),
    message = parsedData.params.message;
  tokenReplacer.replaceTokens(message, process.argv[2], function (error, newMessage) {
    if (error) {
      console.error(error);
      process.exit(1);
    }
    parsedData.params.message = newMessage;
    new HipChatClient().run(parsedData.source, parsedData.params);
  });
});


process.on('uncaughtException', (error) => {
  console.error("An unexpected error occured ", error);
});

scripts/gulpfile.js
// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

var argv = require('yargs').argv,
  gulp = require('gulp'),
  debug = require('gulp-debug'),
  diff = require('gulp-diff'),
  gutil = require('gulp-util'),
  filter = require('gulp-filter'),
  eslint = require('gulp-eslint'),
  beautify = require('gulp-jsbeautifier');

var dropRoot = '../../release/',
  dropFolder = 'workers',
  sourceLocation = '.',
  dropLocation = dropRoot + dropFolder;


// Task: lint-and-beautify
// Performs in-place linting & beautification of the *src* files. We perform
// this in-place so that the checked in code is clean and consistent.
gulp.task('lint-and-beautify', function () {
  // Beautify Config: src/.jsbeautifyrc
  // Linting Config: src/.estlintrc
  return gulp
    .src(['**/*.js', '!node_modules/**/*.js'])
    .pipe(beautify({
      config: '.jsbeautifyrc'
    }))
    .pipe(diff())
    .pipe(diff.reporter({
      quiet: !argv['fail-on-beautify'],
      fail: argv['fail-on-beautify']
    }))
    .pipe(eslint.format())
    .pipe(eslint.failAfterError())
    .pipe(gulp.dest('.'));
});

scripts/tokenReplacement.js
// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var async = require('async'),
  fs = require('fs'),
  url = require('url');

module.exports = {
  replaceTokens: function (message, rootDir, done) {

    if (typeof message === "string") {
      done(null, message);
      return;
    } else if (!message.template) {
      console.error('message param must be a string or contain a template property');
      process.exit(1);
    }

    var buildTokens = {
      "${BUILD_ID}": process.env.BUILD_ID,
      "${BUILD_NAME}": process.env.BUILD_NAME,
      "${BUILD_JOB_NAME}": process.env.BUILD_JOB_NAME,
      "${BUILD_PIPELINE_NAME}": process.env.BUILD_PIPELINE_NAME,
      "${ATC_EXTERNAL_URL}": process.env.ATC_EXTERNAL_URL
    };

    for (var t in buildTokens) {
      message.template = message.template.replaceAll(t, buildTokens[t]);
    }

    if (!message.params || message.params.length === 0) {
      done(null, message.template);
      return;
    }

    var tokenParams = {};

    async.forEachOf(message.params,
      function iterator(item, key, iteratorCallback) {
        if (isFilePath(item)) {
          readFileContents(rootDir, item, function (err, result) {
            if (err) {
              console.error(err);
              process.exit(1);
            }
            tokenParams[key] = result;
            iteratorCallback(null);
          });
        } else {
          tokenParams[key] = item
          iteratorCallback(null);
        }
      },
      function complete(err) {
        if (err) {
          console.error(err);
          process.exit(1);
        }
        for (var t in tokenParams) {
          message.template = message.template.replaceAll('${' + t + '}', tokenParams[t]);
        }

        done(null, message.template);
      }
    );
  }
};

String.prototype.replaceAll = function (search, replacement) {
  var self = this;
  return self.split(search).join(replacement);
};

function isFilePath(value) {
  var parsedPath = url.parse(value);
  return parsedPath.protocol === 'file:';
}

function readFileContents(rootDir, filePath, done) {
  var parsedUrl = url.parse(filePath),
    pathName = rootDir + '/' + parsedUrl.hostname + parsedUrl.pathname;
  async.waterfall([
    function checkIfFileExists(callback) {
      fs.exists(pathName, function (exists) {
        if (exists) {
          return callback(null, pathName);
        }

        console.error("file at path [" + pathName + "] does not exist");
        process.exit(1);
      });
    },
    function readFileText(filePath, callback) {
      return fs.readFile(filePath, 'utf8', callback)
    }
  ], done);
};

scripts/hipChatClient.js
// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var request = require('request');

function HipChatClient() {
  this.sourceProperties = [
    {
      name: "hipchat_server_url",
      optional: false
    },
    {
      name: "token",
      optional: false
    },
    {
      name: "room_id",
      optional: true
    }
  ];
  this.paramProperties = [
    {
      name: "color",
      optional: true
    },
    {
      name: "from",
      optional: false
    },
    {
      name: "message",
      optional: false
    },
    {
      name: "message_format",
      optional: true
    },
    {
      name: "notify",
      optional: true
    }
  ];
  this.validInput = true;
  this.failOnError = true;
}

HipChatClient.prototype.checkArgument = function (argumentName, argumentValue) {
  if (!argumentValue || argumentValue.length === 0) {
    console.error("Please provide a value for", argumentName);
    this.validInput = false;
  }
};

HipChatClient.prototype.sendMessage = function (source, params, done) {
  var requestUrl = `${source.hipchat_server_url}/v2/room/${source.room_id}/notification?auth_token=${source.token}`,
    hipChatMessage = {
      room_id: source.room_id,
      from: params.from,
      message: params.message
    },
    requestOptions = {
      url: requestUrl,
      method: "POST",
      json: hipChatMessage
    };

  if (params.color) {
    hipChatMessage.color = params.color;
  }

  if (params.message_format) {
    hipChatMessage.message_format = params.message_format;
  }

  if (params.notify) {
    hipChatMessage.notify = params.notify;
  }

  request(requestOptions, (err, response) => {
    if (err || response.statusCode > 200) {
      return done(err || response.body)
    }

    return done(err);
  });
};

HipChatClient.prototype.checkProperties = function (values, properties) {
  for (property of properties) {
    if (!property.optional) {
      this.checkArgument(property.name, values[property.name]);
    }
  }
};

HipChatClient.prototype.run = function (source, params) {
  var self = this;
  if (source.fail_on_error === undefined) {
    source.fail_on_error = true;
  }

  self.failOnError = source.fail_on_error;
  self.validInput = true;

  self.checkProperties(source, this.sourceProperties);
  self.checkProperties(params, this.paramProperties);

  if (!source.room_id) {
    var re = /((?:http|https):\/\/.*)\/v2\/room\/(\d+)/;
    if (re.test(source.hipchat_server_url)) {
      var matches = source.hipchat_server_url.match(re);
      source.hipchat_server_url = matches[1];
      source.room_id = parseInt(matches[2], 10);
    } else {
      self.validInput = false;
      console.error("Room id must be included in hipchat_server_url when not specified in the source (https://api.hipchat.com/v2/room/12456)");
    }
  }

  if (!self.validInput) {
    console.error("Please provide valid input and try again");
    return process.exit(1);
  }

  self.sendMessage(source, params, (error, result) => {
    if (error) {
      console.error(`Error sending notification. Fail on error: ${self.failOnError}`);
      console.error(error);
      if (self.failOnError) {
        process.exit(1);
      }
    }

    // Concourse expects this output from stdout, do not use console.dir
    console.log(JSON.stringify({
      version: {
        ref: "none"
      }
    }));
    process.exit(0);
  });
}

module.exports = HipChatClient;